gms/gossiper.cc

```diff
@@ -802,6 +757,84 @@ future<gossiper::endpoint_permit> gossiper::lock_endpoint(inet_address ep) {
     });
 }
 
+
+future<> gossiper::failure_detector_loop_for_node(gms::inet_address node) {
+    auto last = gossiper::clk::now();
+    auto max_duration = std::chrono::milliseconds(_cfg.failure_detector_timeout_in_ms());
+    auto diff = std::chrono::milliseconds(0);
+    auto nr = _live_endpoints.size();
+    int64_t generation = endpoint_state_map[get_broadcast_address()].get_heart_beat_state().get_generation();
+    while (is_enabled()) {
+        bool failed = false;
+        try {
+            logger.debug("failure_detector_loop: Send echo to node {}", node);
+            co_await _messaging.send_gossip_echo(netw::msg_addr(node), generation, max_duration);
+            logger.debug("failure_detector_loop: Send echo to node {}, status = ok", node);
+        } catch (...) {
+            failed = true;
+            logger.info("failure_detector_loop: Send echo to node {}, status = failed: {}", node, std::current_exception());
+        }
+        auto now = gossiper::clk::now();
+        diff = now - last;
+        if (!failed) {
+            last = now;
+        }
+        if (diff > max_duration) {
+            logger.info("failure_detector_loop: Mark node {} as DOWN", node);
+            co_await container().invoke_on(0, [node] (gms::gossiper& g) {
+                return seastar::async([node, &g] {
+                    g.convict(node);
+                });
+            });
+            co_return;
+        } else if (_live_endpoints.size() != nr) {
+            logger.debug("failure_detector_loop: Finish loop for node {}, live_endpoints={}", node, _live_endpoints);
```

Comment:
Why do we exit if the number of endpoints changed?

Reply:
We want to redistribute the nodes each nodes to work on. Each shard is responsible for part of the live nodes.